### PR TITLE
Fix block hash table

### PR DESCRIPTION
**Pull request by plafer**
Fixes the block hash table.

Note that the block stack table also seems broken (which is why I didn't add a `debug_assert_eq!` statement for `p1`). I will fix it in a subsequent PR.

### CODE REVIEW

File: docs/src/design/decoder/constraints.md [hunk 1]
```diff
@@ -287,10 +287,10 @@ Graphically, this looks like so:
 In a similar manner, we define a value representing the result of hash computation as follows:
 
 $$
-bh = \alpha_0 + \alpha_1 \cdot a + \sum_{i=0}^3(\alpha_{i+2} \cdot h_i) + \alpha_7 \cdot h_4 \text{ | degree} = 1
+bh = \alpha_0 + \alpha_1 \cdot a' + \sum_{i=0}^3(\alpha_{i+2} \cdot h_i) + \alpha_7 \cdot f_{is\_loop\_body} \text{ | degree} = 1
 $$
 
-Note that in the above we use $a$ (block address from the current row) rather than $a'$ (block address from the next row) as we did for for values of $ch_1$ and $ch_2$. Also, note that we are not adding a flag indicating whether the block is the first child of a join block (i.e., $\alpha_6$ term is missing). It will be added later on.
+Above, $f_{is\_loop\_body}$ refers to the value in the `IS_LOOP_BODY` column (already constrained to be 0 or 1). Also, note that we are not adding a flag indicating whether the block is the first child of a join block (i.e., $\alpha_6$ term is missing). It will be added later on.
```
Comment (bobbinth) [inline docs/src/design/decoder/constraints.md:293]: Should we mentioned that `f_{is_loop_body}` is actually located in the column `h4` (i.e., `f0` in the illustration)?

File: docs/src/design/decoder/constraints.md [hunk 2]
```diff
@@ -326,17 +326,23 @@ $$
 v_{dyn} = f_{dyn} \cdot ch_{dyn}  \text{ | degree} = 6
 $$
 
+When the `CALL` or `SYSCALL` operation is executed, the hash of the callee is added to the block hash table.
+
+$$
+v_{call\_or\_syscall} = (f_{call} + f_{syscall}) \cdot ch_1  \text{ | degree} = 5
+$$
+
 When `END` operation is executed, hash of the completed block is removed from the block hash table. However, we also need to differentiate between removing the first and the second child of a *join* block. We do this by looking at the next operation. Specifically, if the next operation is neither `END` nor `REPEAT` we know that another block is about to be executed, and thus, we have just finished executing the first child of a *join* block. Thus, if the next operation is neither `END` nor `REPEAT` we need to set the term for $\alpha_6$ coefficient to $1$ as shown below:
```
Comment (bobbinth) [inline docs/src/design/decoder/constraints.md:335]: We also need to update this paragraph to mention the `HALT` operation there.

File: docs/src/design/decoder/constraints.md [hunk 3]
```diff
@@ -326,17 +326,23 @@ $$
...
-u_{end} = f_{end} \cdot (bh + \alpha_6 \cdot (1 - (f_{end}' + f_{repeat}'))) \text{ | } \text{degree} = 8
+u_{end} = f_{end} \cdot (bh + \alpha_6 \cdot (1 - (f_{end}' + f_{repeat}' + f_{halt}'))) \text{ | } \text{degree} = 8
 $$
 
 Using the above definitions, we can describe the constraint for updating the block hash table as follows:
 
 > $$
 p_2' \cdot (u_{end} + 1 - f_{end}) = \\
-p_2 \cdot (v_{join} + v_{split} + v_{loop} + v_{repeat} + v_{dyn} + 1 - (f_{join} + f_{split} + f_{loop} + f_{repeat} + f_{dyn}))
+p_2 \cdot (v_{join} + v_{split} + v_{loop} + v_{repeat} + v_{dyn} + v_{call\_or\_syscall} + 1 - (f_{join} + f_{split} + f_{loop} + f_{repeat} + f_{dyn} + f_{call} + f_{syscall}))
```
Comment (bobbinth) [inline docs/src/design/decoder/constraints.md:345]: Are the $+ f_{call} + f_{syscall}$ at the end of this equation correct? Wouldn't these be redundant with `+ v_{call_or_syscall}`?

**Review by bobbinth — COMMENTED** [review]
Looks good! Thank you for fixing these! I left a couple of questions/comments inline.

File: docs/src/design/decoder/constraints.md (hunk 3 above)
Comment (plafer) [reply docs/src/design/decoder/constraints.md:345]: I think so; it's the same pattern as the other ones. `v_{call_or_syscall}` is for setting `ch1` when $f_{call} + f_{syscall} = 1$, while the $+ f_{call} + f_{syscall}$ is in the `1 - ...` term to preserve the current value of `p2` when $f_{call} + f_{syscall} = 0$

File: docs/src/design/decoder/constraints.md (hunk 1 above)
Comment (plafer) [reply docs/src/design/decoder/constraints.md]: Fixed

File: docs/src/design/decoder/constraints.md (hunk 2 above)
Comment (plafer) [reply docs/src/design/decoder/constraints.md]: fixed

File: docs/src/design/decoder/constraints.md (hunk 3 above)
Comment (bobbinth) [reply docs/src/design/decoder/constraints.md:345]: Ah yes - agreed!

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you!

**Pull request merged**